src/microbe_stage/MicrobeAI.cs

```diff
@@ -370,43 +384,86 @@ private void RunAndTumble(Random random)
+            if (previouslyAbsorbedCompounds.ContainsKey(compoundWeight.Key))
+                compoundDifference -= previouslyAbsorbedCompounds[compoundWeight.Key];
+
+            compoundDifference *= compoundWeight.Value;
+            gradientValue += compoundDifference;
         }
 
+        // Implement a detection threshold to possibly rule out too tiny variations
+        float differenceDetectionThreshold = 0;
+
         // If food density is going down, back up and see if there's some more
-        if (compoundDifference < 0 && random.Next(0, 10) < 9)
+        if (gradientValue < -differenceDetectionThreshold && random.Next(0, 10) < 9)
         {
             MoveWithRandomTurn(2.5f, 3.0f, random);
         }
 
         // If there isn't any food here, it's a good idea to keep moving
-        if (compoundDifference == 0 && random.Next(0, 10) < 5)
+        if (Math.Abs(gradientValue) <= differenceDetectionThreshold && random.Next(0, 10) < 5)
         {
             MoveWithRandomTurn(0.0f, 0.4f, random);
         }
 
-        // If positive last step you gained compounds, so let's stick around
-        if (compoundDifference > 0)
+        // If positive last step you gained compounds, so let's move toward the source
+        if (gradientValue > differenceDetectionThreshold)
         {
-            // There's a decent chance to turn most of the way around
+            // There's a decent chance to turn by 90° to explore gradient
+            // 180° is useless since previous position let you absorb less compounds already
             if (random.Next(0, 10) < 4)
             {
-                MoveWithRandomTurn(0.0f, 3.0f, random);
+                MoveWithRandomTurn(0.0f, 1.5f, random);
             }
         }
     }
 
+    /// <summary>
+    ///   Prioritizing compounds that are stored in lesser quantities.
+    ///   If ATP-producing compounds are low (less than half storage capacities),
+    ///   non ATP-related compounds are discarded.
+    /// </summary>
+    /// <returns> A dictionary of weights for each (useful) compound, to be used in search algorithms.</returns>
+    private Dictionary<Compound, float> ComputeCompoundsSearchWeights(
+        List<KeyValuePair<Compound, float>> usefulCompounds)
+    {
+        // Vital compounds are *direct* ATP producers
+        var usefulVitalCompounds = usefulCompounds.Where(x => x.Key == glucose || x.Key == iron);
+
+        // If this microbe lacks vital compounds don't bother with ammonia and phosphorous
+        // This algorithm doesn't try to determine if iron and hydrogen sulfide is useful to this microbe
+        if (usefulVitalCompounds.Any(
+            compound => microbe.Compounds.GetCompoundAmount(compound.Key) < 0.5f * microbe.Compounds.Capacity))
+        {
+            usefulCompounds = usefulCompounds.Where(x => x.Key != ammonia && x.Key != phosphates).ToList();
+        }
+
+        var compoundsPriority = new Dictionary<Compound, float>();
```

Comment:
I think this dictionary should be stored in the microbe AI class (doesn't need saving) so that new dictionaries don't need to be allocated each update.